src/preferences/pages/blacklist.ts

```diff
@@ -1,111 +1,73 @@
-
-/** Black list Preferences Page */
 export const BlackList = GObject.registerClass(
     {
         Template: uri(import.meta.url, 'blacklist.ui'),
-        GTypeName: 'RoundedWindowCornersPrefsBlacklist',
-        InternalChildren: ['black_list_group', 'add_row_btn'],
+        GTypeName: 'PrefsBlacklist',
+        InternalChildren: ['blacklist_group'],
     },
-    class extends Gtk.Box {
-        private declare _black_list_group: Gtk.ListBox;
-        private declare _add_row_btn: Gtk.Button;
+    class extends Adw.PreferencesPage {
+        private declare _blacklist_group: Adw.PreferencesGroup;
 
         /** Store value of settings */
-        private declare black_list: string[];
+        private declare blacklist: string[];
 
-        _init() {
-            super._init();
-            this.black_list = settings().black_list;
+        constructor() {
+            super();
+            this.blacklist = settings().black_list;
 
-            // Read blacklist from settings, and add it to this._black_list
-            for (const name of this.black_list) {
-                this.on_add_row(name);
+            for (const title of this.blacklist) {
+                this.add_window(undefined, title);
             }
-
-            connections.get().connect(this._add_row_btn, 'clicked', () => {
-                if (this.black_list.includes('')) {
-                    return;
-                }
-
-                this.on_add_row();
-                this.black_list.push('');
-                settings().black_list = this.black_list;
-            });
-
-            connections
-                .get()
-                .connect(
-                    this._black_list_group,
-                    'row-activated',
-                    (_me: Gtk.ListBox, row: Gtk.ListBoxRow) => {
-                        if (row instanceof AppRow) {
-                            row.on_expanded_changed();
-                        }
-                    },
-                );
-        }
-
-        private show_err_msg(item: string) {
-            const title = `${item}: ${_(
-                "Can't add to list, because it has exists",
-            )}`;
-            show_err_msg(title);
         }
 
-        // --------------------------------------------------- [signal handlers]
-
-        private on_delete_row(row: Gtk.ListBoxRow, title: string) {
-            this.black_list.splice(this.black_list.indexOf(title), 1);
-            settings().black_list = this.black_list;
-            this._black_list_group.remove(row);
-        }
-
-        /** Add a row to black list */
-        private on_add_row(title = '') {
-            const handlers: AppRowHandler = {
-                on_delete: (row, title) => this.on_delete_row(row, title),
-                on_title_changed: (old_title, new_title) =>
-                    this.on_title_changed(old_title, new_title),
+        private add_window(_?: Gtk.Button, title?: string) {
```

Comment:
What's the point of the `_?: Gtk.Button` parameter here?

Reply:
This function used both as template callback and as regular function depending on the context. When used as template callback, the source object which issued a signal passes itself to the function. Since we don't really need this button, it simply sits here. But I also wanted to use this function in constructor, so I wouldn't need to repeat same functionality twice, this is why it can be undefined.